### PR TITLE
Add Conditional Order factory

### PR DESCRIPTION
**Pull request by anxolin**
This PR prepares the polling to hook it up to https://github.com/cowprotocol/tenderly-watch-tower

## Conditional Order Factory
The main addition in this PR is this factory.

The factory responsibility is to create Conditional Orders from the `ConditionalOrderParams` (handler, salt, staticInput)

It has some config with the factory methods per handler. 

Some basic usage is:

```ts
# Create factory
const factory = new ConditionalOrderFactory({
  // Register of all known order types
  [TWAP_ADDRESS]: (params) => Twap.fromParams(params),
})

# Then you can create an order just by 
const conditionalOrder = fromParams.fromParams({
   handler,
   salt
   staticInput
})

# Default Registry
The factory requires a registry in order to work
The registry has the different factory methods per handler

The SDK will also export some default config with all the handlers/facttory it knows: for now, only `Twap`: https://github.com/cowprotocol/cow-sdk/pull/151/files#diff-cc93eafdc1844adfa999909a952edcc111abb3031a0eac57403966f6a3c8d00eR5

# Use the order
console.log('Order: ', conditionalOrder.id)

# Now is easy to do polling (and custom validation), without even knowing what order we are processing
console.log('Polling result: ', await order.poll(owner, chainId, provider ))
```

## Companion PR
This PR will have a companion PR in tenderly watch tower project:
https://github.com/cowprotocol/tenderly-watch-tower/pull/16

### CODE REVIEW

**Review by mfw78 — COMMENTED** [review]
Please include open nits from the [review here](https://github.com/cowprotocol/cow-sdk/pull/149#pullrequestreview-1593653332).

**Review by mfw78 — APPROVED** [review]
Looks good, please see nits in the [review here](https://github.com/cowprotocol/cow-sdk/pull/149#pullrequestreview-1593653332) yet to be included (typos etc).

File: src/composable/orderTypes/Twap.ts [hunk 1]
```diff
@@ -179,6 +186,16 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
...
   }
 
+  /**
+   * Create a TWAP order with sound defaults.
+   * @param data The TWAP order parameters in a more user-friendly format.
+   * @returns An instance of the TWAP order.
+   */
+  static fromParams(params: ConditionalOrderParams): Twap {
+    const twap = Twap.deserialize(encodeParams(params))
+    return twap
```
Comment (alfetopito) [inline src/composable/orderTypes/Twap.ts:196]: Nitpick: 

```suggestion
    return Twap.deserialize(encodeParams(params))
```

File: src/composable/orderTypes/Twap.ts [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
+  ConditionalOrderArguments,
+  ConditionalOrderParams,
+  ContextFactory,
+  IsNotValid,
+  IsValid,
+  PollResultErrors,
+} from '../types'
 import { encodeParams, isValidAbi } from '../utils'
-import { SupportedChainId } from 'src/common'
+import { SupportedChainId } from '../../common'
```
Comment (alfetopito) [inline src/composable/orderTypes/Twap.ts:13]: Why changing to relative import?

File: src/composable/orderTypes/index.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { ConditionalOrderRegistry } from '../ConditionalOrderFactory'
+import { TWAP_ADDRESS, Twap } from './Twap'
+export * from './Twap'
+
+export const DEFAULT_CONDITIONAL_ORDER_REGSTRY: ConditionalOrderRegistry = {
+  // Register of all known order types
```
Comment (alfetopito) [inline src/composable/orderTypes/index.ts:6]: Typo?

```suggestion
  // Registry of all known order types
```

**Review by alfetopito — APPROVED** [review]

File: src/composable/orderTypes/Twap.ts (hunk 2 above)
Comment (anxolin) [reply src/composable/orderTypes/Twap.ts:13]: case it was not working in the test

**Pull request merged**